### ISSUE TITLE
How will Doulevo support static web sites?

### ISSUE THREAD

**Issue by ashleydavis** [doulevo/doulevo#18]
Currently Doulevo supports backend services, but it might also be useful for it to support creating, testing and deployment of front ends.

So running `doulevo create --project=react` would create a hello world react project.

It should be possible to run the web site locally using `doulevo run`. This can probably work just out of Docker which means the developer doesn't need to have Node.js or any web server installed.

But how would deployment work?

We shouldn't make any assumption about how the user might want to serve their web page. They might want to deploy it to an AWS S3 bucket or an Azure "Web App".

Normally a Doulevo service would be deployed through `doulevo deploy` which currently builds the Docker image, then generates Kubernetes config file, then runs Kubectl to deploy a container for the service. But for static web sites this would work differently. If the deployment process was "plugable" we could use an AWS or Azure static web site deployment plugin instead of the default Kubenetes deployment plugin.

**Comment by R4meau** [comment]
What if plugins had a second set of questions, just like the one in `template.json`, but for deployment (maybe `deploy.json` in the same dir ad `plugin.json`)?

The plugin would then pick which Dockerfile to run for each supported deployment mechanism.

**Comment by ashleydavis** [comment]
I'm not sure yet. 

I'm thinking the first time a user deploys an application it will ask them what deployment plugin to use. For example they could have many choices such as "Kubernetes", "Azure Container Service", "AWS Elastic Container Service", "Azure Web Site", "AWS Static Site", "Digital Ocean App" etc.

So we ask them on first deployment and then store the data under the project's .doulevo configuration sub-directory.

I think after the we finish (#8) and do the next tech demo we should prototype this to support another deployment target (probably for deploying a static web site).

**Comment by R4meau** [comment]
Oh okay. I actually prefer this better. So we would have two types of plugins:

- creation/template plugins
- deployment plugins

**Comment by ashleydavis** [comment]
Yeah that was always the plan, ultimately Docker also has to be a plugin (e.g. the build/publish plugin).

https://www.the-data-wrangler.com/simplifying-application-deployment-with-doulevo#plugin-based-extensibility

**Comment by R4meau** [comment]
Oh, I see. I missed this from the article.

**Comment by ashleydavis** [comment]
Definitely read it and watch the video as well! You can see my original thinking as well.
